samza-api/src/main/java/org/apache/samza/container/placement/ContainerPlacementMessage.java

```diff
@@ -0,0 +1,152 @@
+
+  /**
+   * Unique identifier of a request or response message. Helps in identifying duplicate request messages.
+   */
+  protected final UUID uuid;
+  /**
+   * Unique identifier for a deployment so messages can be invalidated across a job restarts
+   * for ex yarn bases cluster manager should set this to app attempt id
+   */
+  protected final String deploymentId;
+  // Logical container Id 0, 1, 2
+  protected final String processorId;
+  // Destination host where container is desired to be moved
+  protected final String destinationHost;
+  // Optional request expiry which acts as a timeout for any resource request to cluster resource manager
+  protected final Duration requestExpiry;
+  // Status of the current request
+  protected final StatusCode statusCode;
+  // Timestamp of the request or response message
+  protected final long timestamp;
+
+  protected ContainerPlacementMessage(UUID uuid, String deploymentId, String processorId, String destinationHost,
+      Duration requestExpiry, StatusCode statusCode, long timestamp) {
+    Preconditions.checkNotNull(uuid, "uuid cannot be null");
+    Preconditions.checkNotNull(deploymentId, "deploymentId cannot be null");
+    Preconditions.checkNotNull(processorId, "processorId cannot be null");
+    Preconditions.checkNotNull(destinationHost, "destinationHost cannot be null");
+    Preconditions.checkNotNull(statusCode, "statusCode cannot be null");
+    Preconditions.checkNotNull(timestamp, "Timestamp of the message cannot be null");
+    this.uuid = uuid;
+    this.deploymentId = deploymentId;
+    this.processorId = processorId;
+    this.destinationHost = destinationHost;
+    this.requestExpiry = requestExpiry;
+    this.statusCode = statusCode;
+    this.timestamp = timestamp;
+  }
+
+  protected ContainerPlacementMessage(UUID uuid, String deploymentId, String processorId, String destinationHost,
+      StatusCode statusCode, long timestamp) {
+    this(uuid, deploymentId, processorId, destinationHost, null, statusCode, timestamp);
+  }
+
+  public String getDeploymentId() {
+    return deploymentId;
+  }
+
+  public String getProcessorId() {
+    return processorId;
+  }
+
+  public String getDestinationHost() {
+    return destinationHost;
+  }
+
+  public StatusCode getStatusCode() {
+    return statusCode;
+  }
+
+  public Duration getRequestExpiry() {
+    return requestExpiry;
+  }
+
+  public UUID getUuid() {
+    return uuid;
+  }
+
+  public long getTimestamp() {
+    return timestamp;
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+    ContainerPlacementMessage message = (ContainerPlacementMessage) o;
+    return getUuid().equals(message.getUuid()) && getDeploymentId().equals(message.getDeploymentId())
+        && getProcessorId().equals(message.getProcessorId()) && getDestinationHost().equals(
+        message.getDestinationHost());
```

Comment:
is it expected that we are ignoring requestExpiry, statusCode and timestamp when comparing? Same for other classes.

Reply:
I understand that in reality, each message will have its own UUID, but in terms of comparing java object itself, should we compare all fields? It may be very confusing if two ContainerPlacementMessage equals but their status code is different.

Reply:
I see your point will change it

Reply:
+1, Please make sure all POJO fields are in the equals and hashCode if they can be.